### ISSUE TITLE
Document how to generate and bring your own certs

### ISSUE THREAD

**Issue by mumoshu** [kubernetes-incubator/kube-aws#466]
Extracted from https://github.com/kubernetes-incubator/kube-aws/issues/285

**Comment by zmt** [comment]
Any motion on this. I am going to be forced to abandon this tool unless I can move forward without the private key of the issuing CA on disk to do `kube-aws up`.

**Comment by mumoshu** [comment]
Hi @zmt - Not yet but basically https://github.com/kelseyhightower/kubernetes-the-hard-way/blob/master/docs/02-certificate-authority.md would work.

Btw, is there any other tool that supports this use-case? I wonder which tool you're going to use after abandoning kube-aws. If that works, I can even recommend k8s on aws users who prioritizes this use-case most to use that tool.

**Comment by zmt** [comment]
I was looking at [kops](https://github.com/kubernetes/kops), but the drawback there is elevated IAM permissions required for everyday operations. If I understand correctly, the `kube-aws render credentials` can generate TLS assets for the cluster nodes and the admin and issue from an existing CA. I do not manage the CA infrastructure, so I would need for someone else who has access to a private key to use for this step to do it for me and ship a credentials directory with the generated and issued assets and the CA certificate to trust.

I naively tested simply removing the ca private key from the credentials subdir on a run with a kube-aws generated CA and trying to `kube-aws up --export` where I hit the same error message as in #839 before engaging the team that manages PKI.

**Comment by mumoshu** [comment]
@zmt Thanks for the reply 👍 

Could you elaborate a bit more about the "elevated IAM permissions"?
Perhaps you meant that it requires the single IAM role for every worker/etcd/controller node which is too permissive in respect to which permissions are exactly required for each node?
if that's the case, yes - kube-aws creates or even allows you to bring your own seperate iam role for each type of nodes.

> If I understand correctly, the kube-aws render credentials can generate TLS assets for the cluster nodes and the admin and issue from an existing CA

Yes, you are correct 👍 

For #839, sorry for the inconvenience!
It certainly seems like a bug which is going to be fixed via #877. Btw, I've replied to you there about the issue you encountered.
Once the fix is landed, there will be no problem removing the private key from the credentials directory, as long as tls bootstrapping is disabled(=default).

**Comment by zmt** [comment]
> Could you elaborate a bit more about the "elevated IAM permissions"?

Mainly, I mean creation of KMS and IAM resources for turning up a cluster in a region; to a lesser degree "elevated" permissions could be extended to VPC and networking resource creation. Those permissions are tightly controlled in my org and are not needed for day-to-day operations. Once created, they can be referenced in the ASG creation, etc. Basically, I am trying to iteratively generate different sets of cloudformation json with kube-aws. I requested kms resources so I could then generate cloudformation json from which I am trying to extract just the IAM bits to submit a request with valid cloudformation json for those resources from the same team that manages those things. In parallel, I am waiting for manual creation of the VPC, routing table, and any subnets from a different team because my IAM role does not permit that. Only with all those bits in place, I can generate the cloudformation for just the ASG down and create/destroy just those resources with the permissions in my IAM role. We can't be the only shop that restricts KMS and IAM or separates duties and permissions to different teams/departments. I can see security groups being restricted to a security org also, but that is not true in my case.

> For #839, sorry for the inconvenience!
> It certainly seems like a bug which is going to be fixed via #877. Btw, I've replied to you there about the issue you encountered.
> 

No worries. I appreciate the rapid response. I was contemplating spelunking the code to try to get a pull request going and run forked. A few months back when I used an older version of kube-aws, I had to do that because of hard-coded SANs in the generated PKI for my POC deployment. Then, I didn't have these perms problems b/c I was temporarily granted godmode in AWS to iterate on the POC. Productionizing is different.

**Comment by redbaron** [comment]
@zmt, once https://github.com/kubernetes-incubator/kube-aws/pull/882 is merged, you'll be able to take cert generation completely offline and provide only .enc files for keys. Those files can even come from another team which seems t be the case with your tighten up environment (frankly speaking somebody doesn't understand what benefits cloud provide and still treats VPC as datacenter slowing you down).

With https://github.com/kubernetes-incubator/kube-aws/pull/885 you'll be able to have separate CA and key for TLS bootstrapped workers. It can be self signed CA as nowhere in your organization you'd be connecting to kube cluster using it, so it is completely self contained. As a side effect, there is no need for ca-key.pem at all then and you can simply provide certs pulled from your corporate CA
